src/Memoize.jl

```diff
@@ -75,23 +68,78 @@ macro memoize(args...)
         def_dict[:body] = body
     end
 
+    f = def_dict[:name]
+    sig = :(Tuple{typeof($f), $((splitarg(arg)[2] for arg in def_dict[:args])...)} where {$(def_dict[:whereparams]...)})
+    tail = :(Tuple{$((splitarg(arg)[2] for arg in def_dict[:args])...)} where {$(def_dict[:whereparams]...)})
+
+    scope = gensym()
+    meth = gensym("meth")
+
     esc(quote
-        $Memoize.try_empty_cache($f) # So that redefining a function doesn't leak memory through
-                                     # the previous cache.
         # The `local` qualifier will make this performant even in the global scope.
-        local $fcache = $cache_dict
-        $(cache_name(f)) = $fcache   # for `memoize_cache(f)`
+        local $cache = ($tail, $cache_dict)
+
+        $scope = nothing
+
+        if isdefined($__module__, $(QuoteNode(scope)))
+            function $f end
+
+            # If overwriting a method, empty the old cache.
+            # Notice that methods are hashed by their stored signature
+            try
+                local $meth = which($f, $tail)
+                if $meth.sig == $sig && isdefined($meth.module, :__memories__)
+                    empty!(pop!($meth.module.__memories__, $meth.sig, (nothing, []))[2])
+                end
+            catch
+            end
+        end
+
         $(combinedef(def_dict_unmemoized))
-        Base.@__doc__ $(combinedef(def_dict))
+        local $result = Base.@__doc__($(combinedef(def_dict)))
+
+        if isdefined($__module__, $(QuoteNode(scope)))
+            if !@isdefined __memories__
+                __memories__ = Dict()
+            end
+            # Store the cache so that it can be emptied later
+            local $meth = $which($f, $tail)
+            __memories__[$meth.sig] = $cache
+        end
```

Comment:
Maybe I'm wrong, but this is where I believe that my approach will be just a few extra lines (mostly just to build the cache name with `Symbol(function_name, arg_names...)` instead of ~20.  Macro complexity is especially bad, so unless my suggested approach has a fatal flaw, I believe it will be the way I will take.

Reply:
I suppose our difference of opinion lies in the trade-off between correctness and complexity. I suppose I'll just put my code in a separate package. Thanks for the helpful discussion!